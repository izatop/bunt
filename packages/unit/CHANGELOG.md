# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.6.7](https://github.com/izatop/typesafeunit/compare/v0.6.6...v0.6.7) (2020-03-30)


### Bug Fixes

* logger ([7ca423a](https://github.com/izatop/typesafeunit/commit/7ca423a085a38d524f1a5f0c2545ff95071367ad))





## [0.6.6](https://github.com/izatop/typesafeunit/compare/v0.6.5...v0.6.6) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/unit





## [0.6.5](https://github.com/izatop/typesafeunit/compare/v0.6.4...v0.6.5) (2020-03-30)


### Bug Fixes

* log format ([409f834](https://github.com/izatop/typesafeunit/commit/409f834e428900d4088755fdbfcc2cc1614afaca))
* logger, runtime env ([c269a9e](https://github.com/izatop/typesafeunit/commit/c269a9e1f28f191b05655f4d0d26b9b4ea7624f6))





## [0.6.4](https://github.com/izatop/typesafeunit/compare/v0.6.3...v0.6.4) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/unit





## [0.6.3](https://github.com/izatop/typesafeunit/compare/v0.6.2...v0.6.3) (2020-03-30)


### Bug Fixes

* runtime ([33b4901](https://github.com/izatop/typesafeunit/commit/33b490138dcec7919258e54b4aa1159c8a79df42))





## [0.6.2](https://github.com/izatop/typesafeunit/compare/v0.6.1...v0.6.2) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/unit





## [0.6.1](https://github.com/izatop/typesafeunit/compare/v0.6.0...v0.6.1) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/unit





# [0.6.0](https://github.com/izatop/typesafeunit/compare/v0.5.2...v0.6.0) (2020-03-30)


### Features

* logger, perf etc ([daed7fe](https://github.com/izatop/typesafeunit/commit/daed7fe86ea6e9cb3bd90830a77167e6f9e21755))





## [0.5.2](https://github.com/izatop/typesafeunit/compare/v0.5.1...v0.5.2) (2020-03-26)


### Bug Fixes

* runtime ([4f107b2](https://github.com/izatop/typesafeunit/commit/4f107b2f4dec2e47137f8ebf493756bb478f8480))





## [0.5.1](https://github.com/izatop/typesafeunit/compare/v0.5.0...v0.5.1) (2020-03-26)


### Bug Fixes

* bug ([79eccf3](https://github.com/izatop/typesafeunit/commit/79eccf351511ca30f0d5d66eb7084345fc974a2f))





# [0.5.0](https://github.com/izatop/typesafeunit/compare/v0.4.6...v0.5.0) (2020-03-26)


### Bug Fixes

* bugs ([3d4d640](https://github.com/izatop/typesafeunit/commit/3d4d6405d7e93e5b2c66be8c55015c24c4b7686f))


### Features

* improve flow control ([1d4a31d](https://github.com/izatop/typesafeunit/commit/1d4a31da7b1801fa7eb66e5a7b699a1b79cede2b))





## [0.4.3](https://github.com/izatop/typesafeunit/compare/v0.4.2...v0.4.3) (2020-03-25)


### Bug Fixes

* improve Runtime ([ee6b327](https://github.com/izatop/typesafeunit/commit/ee6b3271278021a3fc3e80cfbe349aa9ce0f656a))





## [0.4.2](https://github.com/izatop/typesafeunit/compare/v0.4.1...v0.4.2) (2020-03-25)


### Bug Fixes

* bugfix ([b0f47b2](https://github.com/izatop/typesafeunit/commit/b0f47b2cf26aa191fc550b5c9fdaf71a0326cdfa))





## [0.4.1](https://github.com/izatop/typesafeunit/compare/v0.4.0...v0.4.1) (2020-03-25)


### Bug Fixes

* imports ([d6d56f4](https://github.com/izatop/typesafeunit/commit/d6d56f4da9cb924083b1e457944ba9b0ee6f2f26))





# [0.4.0](https://github.com/izatop/typesafeunit/compare/v0.3.9...v0.4.0) (2020-03-25)


### Features

* http server ([62abc0f](https://github.com/izatop/typesafeunit/commit/62abc0f20cfefac0a7d5fa6bc91c030719d5e9a8))





## [0.3.9](https://github.com/izatop/typesafeunit/compare/v0.3.8...v0.3.9) (2020-03-08)


### Bug Fixes

* **validation:** skip validation if not required ([d50b893](https://github.com/izatop/typesafeunit/commit/d50b8936a26445b3e098fad01a8c6b0d23933081))





## [0.3.8](https://github.com/izatop/typesafeunit/compare/v0.3.7...v0.3.8) (2020-03-08)


### Bug Fixes

* array validation ([8ba0dda](https://github.com/izatop/typesafeunit/commit/8ba0dda27be76fd68cc519b4a1563f9ca89d04dc))





## [0.3.7](https://github.com/izatop/typesafeunit/compare/v0.3.6...v0.3.7) (2020-03-08)


### Bug Fixes

* array validation ([f271592](https://github.com/izatop/typesafeunit/commit/f27159216ec8985958ef0016b1c175131104b90d))





## [0.3.6](https://github.com/izatop/typesafeunit/compare/v0.3.5...v0.3.6) (2020-03-08)


### Bug Fixes

* error handling ([1792e2b](https://github.com/izatop/typesafeunit/commit/1792e2b83fec230a396d955ab12ccea96e7bc758))





## [0.3.5](https://github.com/izatop/typesafeunit/compare/v0.3.4...v0.3.5) (2020-03-07)

**Note:** Version bump only for package @typesafeunit/unit





## [0.3.4](https://github.com/izatop/typesafeunit/compare/v0.3.3...v0.3.4) (2020-03-07)

**Note:** Version bump only for package @typesafeunit/unit





## [0.3.3](https://github.com/izatop/typesafeunit/compare/v0.3.2...v0.3.3) (2020-03-05)


### Bug Fixes

* **npm:** add LICENSE ([d7bf57f](https://github.com/izatop/typesafeunit/commit/d7bf57fe2dcb9f83cc4a570c99fafcb5e318bdd4))





## [0.3.2](https://github.com/izatop/typesafeunit/compare/v0.3.1...v0.3.2) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/unit





## [0.3.1](https://github.com/izatop/typesafeunit/compare/v0.3.0...v0.3.1) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/unit





# 0.3.0 (2020-03-05)


### Features

* initial commit ([a76a484](https://github.com/izatop/typesafeunit/commit/a76a4849282c382311f4d0cb9f9bd66a10bcceaa))
