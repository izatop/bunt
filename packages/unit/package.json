{
    "name": "@bunt/unit",
    "version": "0.25.0",
    "keywords": [
        "typescript"
    ],
    "author": {
        "name": "Artur Bier",
        "email": "izatop@gmail.com"
    },
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/"
    ],
    "description": "Unit Core",
    "repository": "git@github.com:izatop/bunt.git",
    "scripts": {
        "clean": "yarn build:clean",
        "watch": "yarn build:watch",
        "build": "yarn build:clean && tsc",
        "build:clean": "rimraf dist tsconfig.tsbuildinfo",
        "build:watch": "yarn build --watch"
    },
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@bunt/util": "^0.25.0"
    },
    "license": "MIT"
}
