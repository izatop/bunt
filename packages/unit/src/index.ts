export * from "./interfaces";
export * from "./Unit";
export * from "./Action";
export * from "./Context";
export * from "./Validation";
export * from "./Service";
