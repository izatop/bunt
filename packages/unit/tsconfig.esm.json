{
  "extends": "../../tsconfig.lib.json",
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "Node",
    "rootDir": "src",
    "outDir": "dist/esm"
  },
  "references": [
    {
      "path": "../util/tsconfig.esm.json"
    },
    {
      "path": "../async/tsconfig.esm.json"
    }
  ],
  "include": [
    "src"
  ]
}
