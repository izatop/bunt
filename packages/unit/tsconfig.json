{
    "extends": "../../tsconfig.lib.json",
    "compilerOptions": {
        "rootDir": "src",
        "outDir": "dist"
    },
    "references": [
        {
            "path": "../util"
        }
    ],
    "include": [
        "src"
    ]
}
