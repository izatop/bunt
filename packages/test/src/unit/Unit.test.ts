import {unit, Unit} from "@typesafeunit/unit";
import {BaseTestAction} from "./src/actions/BaseTestAction";
import {NeverRunAction} from "./src/actions/NeverRunAction";
import {ProfileTestAction} from "./src/actions/ProfileTestAction";
import {TestExceptionAction} from "./src/actions/TestExceptionAction";
import {BaseContext} from "./src/context/BaseContext";

test("Unit", async () => {
    const buffer: any[] = [];
    const app = await unit(new BaseContext());
    const profiler = app.getProfiler()
        .listen((type, event) => buffer.push([type, event]));

    expect(await Unit.factory(new BaseContext())).toBeInstanceOf(Unit);
    expect(await Unit.factory(() => new BaseContext())).toBeInstanceOf(Unit);

    await expect(app.run(NeverRunAction)).rejects.toThrow();
    expect(app.has(BaseTestAction)).toBe(false);
    expect(app.add(BaseTestAction)).toEqual([BaseTestAction]);
    expect(app.has(BaseTestAction)).toBe(true);

    const name = Date.now().toString(32);

    const helloWorldRun: string = await app.run(BaseTestAction, {name});
    expect(helloWorldRun).toBe(`Hello, ${name}!`);
    expect(buffer).toMatchSnapshot();
    buffer.length = 0;

    profiler.enable();
    app.add(ProfileTestAction);
    await app.run(ProfileTestAction);
    expect(buffer).toMatchSnapshot();
    buffer.length = 0;

    const error = "Should thrown the Error";
    app.add(TestExceptionAction);
    await expect(app.run(TestExceptionAction, error))
        .rejects
        .toThrow(error);

    expect(buffer).toMatchSnapshot();
    profiler.disable();
});
