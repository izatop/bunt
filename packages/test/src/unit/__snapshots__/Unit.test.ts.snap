// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit 1`] = `Array []`;

exports[`Unit 2`] = `
Array [
  Array [
    "action:validate",
    Object {
      "payload": true,
      "type": "action:validate",
    },
  ],
  Array [
    "action:create",
    Object {
      "payload": undefined,
      "type": "action:create",
    },
  ],
  Array [
    "action:run",
    Object {
      "payload": undefined,
      "type": "action:run",
    },
  ],
  Array [
    "action:success",
    Object {
      "payload": undefined,
      "type": "action:success",
    },
  ],
]
`;

exports[`Unit 3`] = `
Array [
  Array [
    "action:validate",
    Object {
      "payload": true,
      "type": "action:validate",
    },
  ],
  Array [
    "action:create",
    Object {
      "payload": undefined,
      "type": "action:create",
    },
  ],
  Array [
    "action:run",
    Object {
      "payload": "Should thrown the Error",
      "type": "action:run",
    },
  ],
  Array [
    "run:error",
    Object {
      "payload": [Error: Should thrown the Error],
      "type": "run:error",
    },
  ],
  Array [
    "action:fails",
    Object {
      "payload": undefined,
      "type": "action:fails",
    },
  ],
]
`;
